ftprime/argrecorder.py

```diff
@@ -1,48 +1,117 @@
 import msprime
+import numpy as np
 from _msprime import NODE_IS_SAMPLE
-from collections import OrderedDict
 
+NULL_ID = -1
 
-class ARGrecorder(OrderedDict):
+class ARGrecorder(object):
     '''
-    Keys are individual IDs, and values are tuples, whose first entry is a
-    Node, which is a tuple of (birth time, population, is_sample), and the
-    second entry is a ordered list of nonoverlapping Edgesets.  This inherits
-    from OrderedDict so that if individuals are added by birth order then
-    records can easily be output ordered by time.
-
-    Note that the 'time' fields are in *forwards* time, and so when a NodeTable
-    is output for msprime, they must be 'reversed' to be in units of
-    generations ago.
+    To record the ARG, this keeps track of
+        - a NodeTable and an EdgesetTable
+        - information about what has happened since the last simplification
+        - a mapping from (input) individual IDs to (output) node IDs
+    The reason for maintaining this mapping, rather than having (input)
+    individual IDs always equal to the (output) node IDs is to allow periodic
+    simplification, which decouples the two.
+
+    Between simplification steps, the EdgesetTable is appended to, so it is
+    always "up to date".  However, the NodeTable is *not* kept up to date,
+    because its `time` fields are recorded in *time ago*; we also keep track of
+        - a list of birth times of individual IDs
+    which are translated to time-ago at each simplification step, and appended to the Node Table.
+
+    The internal state is stored using
+        - self.node_ids[k] : gives the output Node ID corresponding to the
+          input individual ID k+self.input_min
+    and self.birth_times and self.populations stores the corresponding birth
+    times (in forwards-time) and populations.  These need to contain information
+    about every input individual who might be seen in a new edgeset (ie, who
+    might be a parent).
+
+    Assumes that input individual IDs are *nondecreasing*.
+
+    Must be initialized with a tree sequence ``ts``, this will serve as
+    the history of this first generation of individuals.  The ``samples`` of
+    this tree sequence correspond to the individuals in the first generation.
+
+    If different than the sample IDs of ``ts``, ``first_generation`` has the
+    node IDs of the first generation.
     '''
 
-    def __init__(self):
-        super(ARGrecorder, self).__init__()
-        # need to keep track of this to know where to append new individuals
-        # and how many rows in the NodeTable there are
-        self.num_nodes = 0
-        self.max_time = 0
+    def __init__(self, ts, first_generation=ts.samples(), time=0.0, max_indivs=1000):
```

Comment:
Probably a bit cleaner to have ``first_generation=None`` by default, and then if it is None take it equal to ``ts.samples``. I'm surprised if this is valid Python actually, I thought default values were resolved at import time...